src/poisson/poisson_process.jl

```diff
@@ -0,0 +1,125 @@
+"""
+    PoissonProcess{R,D}
+
+Homogeneous temporal Poisson process with arbitrary mark distribution.
+
+# Fields
+
+- `λ::R`: ground intensity.
+- `mark_dist::D`: mark distribution.
+
+# Constructor
+
+    PoissonProcess(λ, mark_dist)
+"""
+struct PoissonProcess{R<:Real,D} <: AbstractPointProcess
+    λ::R
+    mark_dist::D
+
+    function PoissonProcess(λ::R, mark_dist::D; check_args::Bool=true) where {R<:Real,D}
+        check_args &&
+            λ < zero(λ) &&
+            throw(
+                DomainError(
+                    "λ = $λ", "PoissonProcess: the ground intensity λ must be non negative."
+                ),
+            )
+        return new{R,D}(λ, mark_dist)
+    end
+end
+
+function Base.show(io::IO, pp::PoissonProcess)
+    return print(io, "PoissonProcess($(pp.λ), $(pp.mark_dist))")
+end
+
+## Alias 
+"""
+    UnivariatePoissonProcess{R}
+
+Homogeneous univariate temporal Poisson process with scalar intensity `λ::R`.
+
+`UnivariatePoissonProcess{R}` is simply a type alias for `PoissonProcess{R,Dirac{Nothing}}`.
+"""
+const UnivariatePoissonProcess{R<:Real} = PoissonProcess{R,Dirac{Nothing}}
+
+function Base.show(io::IO, pp::UnivariatePoissonProcess)
+    return print(io, "UnivariatePoissonProcess($(pp.λ))")
+end
+
+"""
+    MultivariatePoissonProcess{R}
+
+Homogeneous multivariate temporal Poisson process with marginal intensities of type `R`.
+
+`MultivariatePoissonProcess{R}` is simply a type alias for `PoissonProcess{R,Categorical{R,Vector{R}}}`.
+"""
+const MultivariatePoissonProcess{R<:Real} = PoissonProcess{R,Categorical{R,Vector{R}}}
```

Comment:
Just realized that the automatic conversion made by `Distributions.jl` makes, for instance, that `PoissonProcess(Int[1,1]) isa MultivariatePoissonProcess{Int}` is false.
If this "problem" happens only in the case of integers intensities, we could just don't care, but it may also arises to other `Real`'s, no ?

Reply:
Distributions.jl has a lot of issues related to types. For my package [HiddenMarkovModels.jl](https://github.com/gdalle/HiddenMarkovModels.jl), I actually rolled out my own version of `Categorical`, only to avoid them. There was a project called [MeasureTheory.jl](https://github.com/JuliaMath/MeasureTheory.jl) to rebuild Distributions.jl on a saner basis, but I think it's dead at the moment.

Either we decide that type issues are not our fault if they come from Distributions.jl, or we fix them ourselves. I'm not sure which is the best approach.

Reply:
A possibility would be to write
```suggestion
const MultivariatePoissonProcess{R<:Real} = PoissonProcess{R,Categorical{T,Vector{T}}} where T
```

Reply:
From what I understand from `Categorical` in `Distributions.jl`, the type parameter is actually associated with the type of the probabilities in the probability vector. So, the type must be an `AbstractFloat`, since it is a number between 0 and 1. Except, for example, in the case `Categorical([1, 0])`, this returns a `Categorical{Int64, Vector{Int64}}`.

If the intensities are integers, the type for `Categorical` will still be an `AbstratFloat`, since it will be the type of the resulting probabilities. What seems to work is
```julia
const MultivariatePoissonProcess{R<:Real} = PoissonProcess{R,Categorical{F,Vector{F}}} where {F<:AbstractFloat}
```

Reply:
> Here is an other suggestion :
> 
> ```julia
> const MultivariatePoissonProcess{R<:Real, F} = PoissonProcess{R,Categorical{F,Vector{F}}}
> ```
> 
> Hence, this is a `DataType`.
> 
> The drawback is that you need to provide e.g. `MultivariatePoissonProcess{Float32, Float32}` instead of `MultivariatePoissonProcess{Float32}` to the `fit` function.

Don't we need to define a separate `fit` method anyway (or, at least, `suffstats`)? From what I understand, `MultivariatePoissonProcess` is for non-marked events, so when we call `fit` with some vector of histories, we cannot use the marks in the histories, we need build the marks so that each event time has a mark corresponding to which history it came from.

If that is the case, couldn't we just define the method `fit(pp::Type{<:MultivariatePoissonProcess{R}}, hists::Vector{<:History}) where {R<:Real}`, instead of using `MultivariatePoissonPoisson{R,F}`? Or have both, but the first just dispatches to the second one with `F = float(R)`?